(BEDS-1368) adjust deposits endpoints to new api structure, pectra changes

remoterami opened the pull request.

remoterami committed:
refactor!: adjust to new api structure, pectra changes

LuccaBitfly commented on backend/pkg/api/data_access/dummy.go:
```diff
@@ -369,19 +369,19 @@ func (*DummyService) GetValidatorDashboardGroupHeatmap(ctx context.Context, dash
 
-func (*DummyService) GetValidatorDashboardTotalElDeposits(ctx context.Context, dashboardId t.VDBId) (*t.VDBTotalExecutionDepositsData, error) {
+func (*DummyService) GetValidatorDashboardTotalElDeposits(ctx context.Context, dashboardId t.VDBId, search string) (*t.VDBTotalExecutionDepositsData, error) {
 	return getDummyStruct[t.VDBTotalExecutionDepositsData](ctx)
 }
 
-func (*DummyService) GetValidatorDashboardTotalClDeposits(ctx context.Context, dashboardId t.VDBId) (*t.VDBTotalConsensusDepositsData, error) {
+func (*DummyService) GetValidatorDashboardTotalClDeposits(ctx context.Context, dashboardId t.VDBId, search string) (*t.VDBTotalConsensusDepositsData, error) {
```
**question:** the total endpoints should be affected by search?

remoterami replied:
I think so, yes, that's what we are currently doing for withdrawals at least https://github.com/gobitfly/beaconchain/blob/staging/backend/pkg/api/data_access/vdb_withdrawals.go#L440-L446

LuccaBitfly commented on backend/pkg/api/handlers/input_validation.go:
```diff
@@ -348,6 +348,24 @@ func (v *validationError) checkPagingParams(q url.Values) Paging {
 	return paging
 }
 
+func (v *validationError) checkPagingMap(params map[string]string) Paging {
```
**note (non-blocking):** down the line this func should totally replace `checkPagingParams` :+1: (feel free to resolve this comment)

LuccaBitfly commented on backend/pkg/api/handlers/validator_dashboard.go:
```diff
@@ -249,3 +249,163 @@ func (h *HandlerService) GetValidatorDashboardSummaryChart(ctx context.Context,
+
+type inputGetValidatorDashboardTotalConsensusLayerDeposits struct {
+	search      string
+	dashboardId interface{}
```
**nitpick (non-blocking):** This should be `dashboardIdParam` and not `dashboardId`, but looking through the code I've been inconsistent with this myself :sweat_smile:

LuccaBitfly commented on backend/pkg/api/types/validator_dashboard.go:
```diff
@@ -240,19 +240,22 @@ type VDBExecutionDepositsTableRow struct {
 	TxHash               Hash            `json:"tx_hash"`
 	WithdrawalCredential Hash            `json:"withdrawal_credential"`
 	Amount               decimal.Decimal `json:"amount"`
-	Valid                bool            `json:"valid"`
+	Validity             string          `json:"validity" tstype:"'valid' | 'invalid' | 'invalid_skipped'" faker:"oneof: valid, invalid, invalid_skipped"`
 }
 type GetValidatorDashboardExecutionLayerDepositsResponse ApiPagingResponse[VDBExecutionDepositsTableRow]
 
 type VDBConsensusDepositsTableRow struct {
 	PublicKey            PubKey          `json:"public_key"`
 	Index                uint64          `json:"index"`
 	GroupId              uint64          `json:"group_id"`
-	Epoch                uint64          `json:"epoch"`
-	Slot                 uint64          `json:"slot"`
+	SlotQueued           uint64          `json:"slot_queued"`
+	SlotProcessed        uint64          `json:"slot_processed"`
 	WithdrawalCredential Hash            `json:"withdrawal_credential"`
 	Amount               decimal.Decimal `json:"amount"`
 	Signature            Hash            `json:"signature"`
+	Type                 string          `json:"type" tstype:"'manual' | 'auto'" faker:"oneof: manual, auto"`
+	Status               string          `json:"status" tstype:"'queued' | 'completed' | 'rejected'" faker:"oneof: queued, completed, rejected"`
+	RejectReason         *string         `json:"reject_reason,omitempty" tstype:"'invalid_signature'" faker:"oneof: invalid_signature"`
```
**suggestion:** You could add more faker fields which the frontend team would appreciate, e.g. `faker:"eth"` to the `decimal.Decimal` fields for proper Wei values already exists. See: https://github.com/gobitfly/beaconchain/blob/538cc1e4d0e92864e20b834c3f7b21ebe61e6f40/backend/pkg/api/data_access/dummy.go#L32-L34 Could also add more custom providers for hashes, signatures etc., up to you really.

remoterami replied:
Thanks for the tip, played around and added a few examples. ptal

LuccaBitfly commented on frontend/types/api/validator_dashboard.ts:
```diff
@@ -207,18 +207,21 @@ export interface VDBExecutionDepositsTableRow {
 export type GetValidatorDashboardExecutionLayerDepositsResponse = ApiPagingResponse<VDBExecutionDepositsTableRow>;
 export interface VDBConsensusDepositsTableRow {
   public_key: PubKey;
   index: number /* uint64 */;
   group_id: number /* uint64 */;
-  epoch: number /* uint64 */;
-  slot: number /* uint64 */;
+  slot_queued: number /* uint64 */;
+  slot_processed: number /* uint64 */;
   withdrawal_credential: Hash;
   amount: string /* decimal.Decimal */;
   signature: Hash;
+  type: 'manual' | 'auto';
+  status: 'queued' | 'completed' | 'rejected';
+  reject_reason?: 'invalid_signature';
 }
```
**nitpick:** Don't know if it's applicable in this case, but frontend could appreciate a union type when there are mutually exclusive fields (assuming it is the case here with `rejection_reason`):
```ts
export type VDBConsensusDepositsTableRow =
  | {
      public_key: PubKey;
      index: number /* uint64 */;
      group_id: number /* uint64 */;
      slot_queued: number /* uint64 */;
      slot_processed: number /* uint64 */;
      withdrawal_credential: Hash;
      amount: string /* decimal.Decimal */;
      signature: Hash;
      type: 'manual' | 'auto';
      status: 'queued' | 'completed';
    }
  | {
      public_key: PubKey;
      index: number /* uint64 */;
      group_id: number /* uint64 */;
      slot_queued: number /* uint64 */;
      slot_processed: number /* uint64 */;
      withdrawal_credential: Hash;
      amount: string /* decimal.Decimal */;
      signature: Hash;
      type: 'manual' | 'auto';
      status: 'rejected';
      rejection_reason: 'invalid_signature';
    };
```
Though this has to be coded manually into the go file, similarly to how it's done [here](https://github.com/gobitfly/beaconchain/blob/538cc1e4d0e92864e20b834c3f7b21ebe61e6f40/backend/pkg/api/types/search.go#L39) to achieve [this](https://github.com/gobitfly/beaconchain/blob/538cc1e4d0e92864e20b834c3f7b21ebe61e6f40/frontend/types/api/search.ts#L34-L36). There is probably a cleaner way to achieve this using the `tygo:emit` directive, (see docs [here](https://github.com/gzuidhof/tygo/blob/main/README.md#custom-type-mapping)) but I haven't put too much thought into it. If you decide to implement this, I'd be happy to help out.

marcel-bitfly replied:
dunno if it would make sense to strip out only the parts that actually are different?

```typescript
{
      public_key: PubKey;
      index: number /* uint64 */;
      group_id: number /* uint64 */;
      slot_queued: number /* uint64 */;
      slot_processed: number /* uint64 */;
      withdrawal_credential: Hash;
      amount: string /* decimal.Decimal */;
      signature: Hash;
      type: 'manual' | 'auto';
    }
  & ( 
       {
         status: 'rejected';
         rejection_reason: 'invalid_signature';
       }
       | {
       status: 'queued' | 'completed';
       }
   );
```

LuccaBitfly replied:
after discussing with FE team, won't do the extra work on our side until NOT having union types gives them more extra work on their side.

LuccaBitfly commented on backend/pkg/api/types/validator_dashboard.go:
```diff
@@ -240,19 +240,22 @@ type VDBExecutionDepositsTableRow struct {
 	WithdrawalCredential Hash            `json:"withdrawal_credential"`
 	Amount               decimal.Decimal `json:"amount"`
-	Valid                bool            `json:"valid"`
+	Validity             string          `json:"validity" tstype:"'valid' | 'invalid' | 'invalid_skipped'" faker:"oneof: valid, invalid, invalid_skipped"`
```
**question:** 
![image](https://github.com/user-attachments/assets/a9e1a317-f28f-40ff-8774-892d9ccf86a9)
This is an instance of `invalid_skipped`?

remoterami replied:
Yes correct.

For context: The signature only has to be correct for a validators first deposit, in all subsequent cases the signature verification is skipped. So even if the signature is technically incorrect in these cases, the deposit will still be included & accepted on chain. `invalid_skipped` was the best validity type naming for this that I came up with

LuccaBitfly commented on frontend/types/api/validator_dashboard.ts:
```diff
@@ -207,18 +207,21 @@ export interface VDBExecutionDepositsTableRow {
   signature: Hash;
+  type: 'manual' | 'auto';
+  status: 'queued' | 'completed' | 'rejected';
+  reject_reason?: 'invalid_signature';
```
**issue:** Is this correct? The reject reason can only be `'invalid_signature'` and nothing else? In this case, the field is redundant, right?

remoterami replied:
> In this case, the field is redundant, right?

Technically yes, but I chose to do it like this for 2 reasons:
- compatability with future changes: similar to how pectra introduced rejection reasons for withdrawals, the same could happen to deposits later
- consistency with the other event table endpoints (withdrawals & consolidations)

LuccaBitfly commented on backend/pkg/api/types/validator_dashboard.go:
```diff
@@ -240,19 +240,22 @@ type VDBExecutionDepositsTableRow struct {
 }
 type GetValidatorDashboardExecutionLayerDepositsResponse ApiPagingResponse[VDBExecutionDepositsTableRow]
 
 type VDBConsensusDepositsTableRow struct {
```
**question:** From CL deposits design:
![image](https://github.com/user-attachments/assets/c2a2db1d-0445-44e7-8ab9-9d6828ddf6bc)
Which field will cause this text to be shown?

remoterami replied:
If `type == auto`; I've added an explainer comment to the designs

LuccaBitfly reviewed (approved):
lgtm

enzo-bitfly commented on backend/pkg/api/enums/validator_dashboard_enums.go:
```diff
@@ -248,6 +248,76 @@ var VDBWithdrawalsColumns = struct {
+	return int(c)
+}
+
+func (VDBDepositsElColumn) NewFromString(s string) VDBDepositsElColumn {
+	switch s {
+	case "", "block":
+		return VDBDepositElBlock
+	case "amount":
+		return VDBDepositElAmount
+	default:
+		return VDBDepositsElColumn(-1)
+	}
+}
```
question: could you add the alias to sort by `timestamp`?

remoterami replied:
right sure, done

enzo-bitfly committed:
refactor: rename <BcTable> prop to `is-loading`

enzo-bitfly committed:
feat: adapt `EL deposits table` for Pectra

Move fetching logic to `dashboard` root page component
See: BEDS-1362

enzo-bitfly committed:
refactor(variables.scss): replace spacing `px` values by `rem`

enzo-bitfly committed:
refactor(useCustomFetch): allow endpoint-specific `is_mocked` param to override global setting

enzo-bitfly committed:
feat: add `BcIconFlag`

enzo-bitfly committed:
feat: adapt `CL deposits table` for Pectra

-Move fetching logic to `dashboard` root page component
-Ensure immediate fetch for active tab on first page load

See: BEDS-1362
